spatial_batch_kernels.hpp

```diff
@@ -0,0 +1,1122 @@
+      const auto warpCount = split::s_pop_count(mask);
+      if (w_tid == 0) {
+         warpSums[wid] = warpCount;
+      }
+      __syncthreads();
+      // Figure out the total here because we overwrite shared mem later
+      if (wid == 0) {
+         // ceil int division
+         int activeWARPS = nextPow2(1 + ((current - 1) / WARPLENGTH));
+         auto reduceCounts = [activeWARPS](int localCount) -> int {
+                                for (int i = activeWARPS / 2; i > 0; i = i / 2) {
+                                   localCount += split::s_shuffle_down(localCount, i, SPLIT_VOTING_MASK);
+                                }
+                                return localCount;
+                             };
+         auto localCount = warpSums[w_tid];
+         int totalCount = reduceCounts(localCount);
+         if (w_tid == 0) {
+            outputCount = totalCount;
+            outputSize += totalCount;
+            assert((outputSize <= capacity) && "extract_GIDs_kernel ran out of capacity!");
+            outputVec->device_resize(outputSize);
+         }
+      }
+      // Prefix scan WarpSums on the first warp
+      if (wid == 0) {
+         auto value = warpSums[w_tid];
+         for (uint d = 1; d < warpsPerBlock; d = 2 * d) {
+            int res = split::s_shuffle_up(value, (int)d, SPLIT_VOTING_MASK);
+            if (tid % warpsPerBlock >= d) {
+               value += res;
+            }
+         }
+         warpSums[w_tid] = value;
+      }
+      __syncthreads();
+      auto offset = (wid == 0) ? 0 : warpSums[wid - 1];
+      auto pp = split::s_pop_count(mask & ((ONE << w_tid) - ONE));
+      const auto warpTidWriteIndex = offset + pp;
+      if (active) {
+         if constexpr (FIRSTONLY) {
+            output[warpTidWriteIndex] = input[tid].first;
+         } else {
+            output[warpTidWriteIndex] = input[tid];
+         }
+      }
+      // Next loop iteration:
+      input += current;
+      output += outputCount;
+      remaining -= current;
+   }
+   __syncthreads();
+   if (tid == 0) {
+      // Resize to final correct output size.
+      outputVec->device_resize(outputSize);
+      output_sizes[hashmapIndex] = outputSize;
+   }
+}
+
+template <typename Rule, typename ELEMENT, bool FIRSTONLY=false>
+void extract_GIDs_kernel_launcher(
+   Hashinator::Hashmap<vmesh::GlobalID,vmesh::LocalID>** input_maps,
+   split::SplitVector<ELEMENT> **output_vecs,
+   vmesh::LocalID* output_sizes,
+   Rule rule,
+   vmesh::VelocityMesh** rule_meshes,
+   Hashinator::Hashmap<vmesh::GlobalID,vmesh::LocalID> **rule_maps,
+   split::SplitVector<vmesh::GlobalID> **rule_vectors,
+   const uint nCells,
+   gpuStream_t stream
+   ) {
+   extract_GIDs_kernel<Rule,ELEMENT,FIRSTONLY><<<nCells, Hashinator::defaults::MAX_BLOCKSIZE, 0, stream>>>(
+      input_maps,
+      output_vecs,
+      output_sizes,
+      rule,
+      rule_meshes,
+      rule_maps,
+      rule_vectors
+      );
+   CHK_ERR( gpuPeekAtLastError() );
+}
+
+/*
```

Comment:
Like many, could be doxygen style, but...

Reply:
Yeah, that would be a nice todo. Perhaps for a summer student :)